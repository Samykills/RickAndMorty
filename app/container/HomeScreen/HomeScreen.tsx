import React from 'react';
import { FlatList, View, Text, StyleSheet, Animated, Alert, Platform } from 'react-native';
import ListItem from '../../component/ListItem';
import RickAndMortyApiService from '../../service/RickAndMortyApiService';
import FAB from '../../component/FAB';
import Header from '../../component/Header';

const HEADER_MAX_HEIGHT = 200;
const NUMBER_OF_COLUMNS = 3;
const INITIAL_PAGE = 1;

class HomeScreen extends React.Component {
  state = {
    scrollY: new Animated.Value(0),
    pageNo: INITIAL_PAGE,
    characterList: [],
    apiInfo: {},
    loading: false,
    totalPages: INITIAL_PAGE
  };

  componentDidMount = () => {
    this.fetchCharacterData();
  };

  fetchCharacterData = () => {
    if (!this.state.loading) {
      this.setState({ loading: true });
      RickAndMortyApiService.fetchCharacters(this.state.pageNo).then(
        (res: any) => {
          this.setState({
            apiInfo: res.info,
            characterList: [].concat(this.state.characterList, res.results),
            pageNo: this.state.pageNo + 1,
            loading: false,
            totalPages: res.info.pages
          });
        },
        err => {
          this.setState({ loading: false });
          Alert.alert('', err.toString());
        }
      );
    }
  };

  _renderScrollViewContent = () => {
    return (
      <FlatList
        scrollEventThrottle={16}
        onScroll={Animated.event([{ nativeEvent: { contentOffset: { y: this.state.scrollY } } }])}
        numColumns={NUMBER_OF_COLUMNS}
        contentContainerStyle={styles.scrollViewContent}
        data={this.state.characterList}
        refreshing={this.state.loading}
        onEndReached={this.fetchCharacterData}
        onEndReachedThreshold={0.99}
        removeClippedSubviews={true}
        renderItem={({ item }) => {
          return <ListItem item={item} numColumns={NUMBER_OF_COLUMNS} />;
        }}
        ListEmptyComponent={() => {
          return (
            <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
              <Text style={{ textAlign: 'center', fontWeight: 'bold' }}>All Characters from the show Rick and Morty!</Text>
            </View>
          );
        }}
        keyExtractor={(item, index) => {
          return index + '';
        }}
      />
    );
  };

  render() {
    return (
      <View style={styles.fill}>
        {this._renderScrollViewContent()}
        <Header scrollY={this.state.scrollY} />
        <FAB scroll={this.state.scrollY} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  fill: {
    flex: 1,
    backgroundColor: '#EEEEEE'
  },
  scrollViewContent: {
    marginTop: HEADER_MAX_HEIGHT
  }
});

export default HomeScreen;
