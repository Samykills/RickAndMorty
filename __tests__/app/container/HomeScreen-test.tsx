// __tests__/Intro-test.js
import React from 'react';
import HomeScreen from '../../../app/container/HomeScreen/HomeScreen';
import renderer from 'react-test-renderer';
beforeEach(() => {
  jest.useFakeTimers();
});
test('Header renders correctly', () => {
  const tree = renderer.create(<HomeScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
