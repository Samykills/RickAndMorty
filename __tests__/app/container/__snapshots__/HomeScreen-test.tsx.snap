// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#EEEEEE",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "marginTop": 200,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.99}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={true}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEventThrottle={16}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
        >
          All Characters from the show Rick and Morty!
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "#212121",
        "height": 200,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": "https://comicsalliance.com/files/2016/01/rick-and-morty.jpg",
        }
      }
      style={
        Object {
          "height": 200,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "resizeMode": "cover",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
          "width": undefined,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "height": 32,
          "justifyContent": "center",
          "marginTop": 38,
          "opacity": 0,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "transparent",
            "color": "white",
            "fontSize": 18,
          }
        }
      >
        All Characters
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 50,
        "justifyContent": "center",
        "opacity": 0,
        "position": "absolute",
        "right": 25,
        "transform": Array [
          Object {
            "translateY": 200,
          },
        ],
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderRadius": 28,
            "height": 52,
            "justifyContent": "center",
            "margin": 7,
            "shadowColor": "rgba(0,0,0, .4)",
            "shadowOffset": Object {
              "height": 1,
              "width": 1,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
            "width": 52,
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
        underlayColor="white"
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#263238",
                "fontSize": 24,
              },
              Object {
                "backgroundColor": "transparent",
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
          testID="iconIcon"
        >
          ï€‚
        </Text>
      </View>
    </View>
  </View>
</View>
`;
